// vike/dist/esm/node/vite/plugins/non-runnable-dev/pluginViteRPC.js
export { pluginViteRPC };
import { createViteRPC, isDevCheck } from '../../utils.js';
import { retrievePageAssetsDev } from '../../../runtime/renderPage/getPageAssets/retrievePageAssetsDev.js';
import { getViteConfigRuntime } from '../../shared/getViteConfigRuntime.js';
function getViteRpcFunctions(viteDevServer) {
    return {
        async transformIndexHtmlRPC(html) {
            return await viteDevServer.transformIndexHtml('/', html);
        },
        async retrievePageAssetsDevRPC(clientDependencies, clientEntries) {
            return await retrievePageAssetsDev(viteDevServer, clientDependencies, clientEntries);
        },
        async getViteConfigRuntimeRPC() {
            return getViteConfigRuntime(viteDevServer.config);
        },
    };
}
function pluginViteRPC() {
    return [
        {
            name: 'vike:pluginViteRPC:1',
            apply: (_, configEnv) => isDevCheck(configEnv),
            configureServer: {
                handler(viteDevServer) {
                    createViteRPC(viteDevServer, getViteRpcFunctions);
                },
            },
        },
    ];
}
